File: leetcode/easy/938-range-sum/bfs.cpp
#include <algorithm>
#include <climits>
#include <deque>
#include <iostream>
#include <list>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
 public:
  int rangeSumBST(TreeNode *root, int low, int high) {
    if (!root) return 0;

    int total = 0;
    vector<TreeNode *> currLevel = {root};

    while (!currLevel.empty()) {
      vector<TreeNode *> nextLevel;
      for (auto &node : currLevel) {
        if (node->val > high) {
          if (node->left) nextLevel.push_back(node->left);
        } else if (node->val < low) {
          if (node->right) nextLevel.push_back(node->right);
        } else {
          total += node->val;
          if (node->left) nextLevel.push_back(node->left);
          if (node->right) nextLevel.push_back(node->right);
        }
      }

      currLevel = nextLevel;
    }

    return total;
  }
};

File: leetcode/easy/938-range-sum/bfs.py
"""
  @author Vortexx2
  Problem 938 - Range Sum of BST
  
  Runtime - 192 ms
  Memory Usage - 22.5 MB
"""
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
    if not root:
      return 0

    currLevel = [root]
    total = 0

    while currLevel:
      nextLevel = []

      for node in currLevel:
        if node.val < low:
          if node.right:
            nextLevel.append(node.right)
          
        elif node.val > high:
          if node.left:
            nextLevel.append(node.left)

        else:
          total += node.val
          if node.left: nextLevel.append(node.left)
          if node.right: nextLevel.append(node.right)

      currLevel = nextLevel

    return total
